modules/generator/instance.go

```diff
@@ -0,0 +1,279 @@
+
+	return i, nil
+}
+
+func (i *instance) watchOverrides() {
+	reloadPeriod := 10 * time.Second
+
+	ticker := time.NewTicker(reloadPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			err := i.updateProcessors(i.overrides.MetricsGeneratorProcessors(i.instanceID))
+			if err != nil {
+				metricActiveProcessorsUpdateFailed.WithLabelValues(i.instanceID).Inc()
+				level.Error(log.Logger).Log("msg", "updating the processors failed", "err", err, "tenant", i.instanceID)
+			}
+
+		case <-i.shutdownCh:
+			return
+		}
+	}
+}
+
+func (i *instance) updateProcessors(desiredProcessors map[string]struct{}) error {
+	i.processorsMtx.RLock()
+	defer i.processorsMtx.RUnlock()
+
+	// add missing processors
+	for processorName := range desiredProcessors {
+		_, ok := i.processors[processorName]
+		if ok {
+			// processor already exists
+			continue
+		}
+
+		var newProcessor processor.Processor
+		switch processorName {
+		case spanmetrics.Name:
+			newProcessor = spanmetrics.New(i.cfg.Processor.SpanMetrics, i.instanceID)
+		case servicegraphs.Name:
+			newProcessor = servicegraphs.New(i.cfg.Processor.ServiceGraphs, i.instanceID)
+		default:
+			level.Error(log.Logger).Log(
+				"msg", fmt.Sprintf("processor does not exist, supported processors: [%s]", strings.Join(allSupportedProcessors, ", ")),
+				"processorName", processorName,
+				"tenant", i.instanceID,
+			)
+			// this is most likely a misconfiguration, abort updateProcessors before we remove any active processors
+			return fmt.Errorf("unknown processor %s", processorName)
+		}
+
+		i.processorsMtx.RUnlock()
+		err := i.addProcessor(newProcessor)
+		i.processorsMtx.RLock()
+
+		if err != nil {
+			return err
+		}
+	}
+
+	// remove processors that are not in desiredProcessors
+	for processorName := range i.processors {
+		_, ok := desiredProcessors[processorName]
+		if ok {
+			// processor is also in the desired list
+			continue
+		}
+
+		i.processorsMtx.RUnlock()
+		i.removeProcessor(processorName)
+		i.processorsMtx.RLock()
+	}
+
+	i.updateProcessorMetrics()
+
+	return nil
+}
+
+func (i *instance) addProcessor(processor processor.Processor) error {
+	level.Debug(log.Logger).Log("msg", "adding processor", "processorName", processor.Name(), "tenant", i.instanceID)
+
+	i.processorsMtx.Lock()
+	defer i.processorsMtx.Unlock()
+
+	// check the processor wasn't added in the meantime
+	if _, ok := i.processors[processor.Name()]; ok {
+		return nil
+	}
+
+	err := processor.RegisterMetrics(i.registry)
+	if err != nil {
+		return fmt.Errorf("error registering metrics for %s: %w", processor.Name(), err)
+	}
+
+	i.processors[processor.Name()] = processor
+
+	return nil
+}
+
+func (i *instance) removeProcessor(processorName string) {
+	i.processorsMtx.Lock()
+	defer i.processorsMtx.Unlock()
+
+	deletedProcessor, ok := i.processors[processorName]
+	if !ok {
+		return
+	}
+
+	delete(i.processors, processorName)
+
+	err := deletedProcessor.Shutdown(context.TODO(), i.registry)
```

Comment:
doing this under lock could be dangerous if shutdown includes flushing any data. can we do this outside the lock?

or maybe just establishing the expectation that `Shutdown` has to be synchronous and fast is enough.

Reply:
Shutting down the processor should not flush any data. This is only meant to release any resources created by the processor (e.g. goroutines). If we want to flush the data before shutdown we should call `i.collectAndPushMetrics` first (like we do in `instance.shutdown`). `processor.Shutdown` will unregister all the counters from the registry, so this data is deleted immediately.

Agree there is no reason to do this while holding the lock, besides that the code is simpler. Once the processor is removed from the map we can release the lock.

Let me rethink this whole lock situation a bit, we should be able to make this a bit simpler / easier to follow 🤔